tls/extensions/s2n_server_key_share.c

```diff
@@ -238,6 +238,21 @@ static int s2n_server_key_share_recv_ecc(struct s2n_connection *conn, uint16_t n
     }
 
     struct s2n_ecc_evp_params *server_ecc_evp_params = &conn->kex_params.server_ecc_evp_params;
+
+    /**
+     *= https://tools.ietf.org/rfc/rfc8446#4.2.8
+     *# If using (EC)DHE key establishment and a HelloRetryRequest containing a
+     *# "key_share" extension was received by the client, the client MUST
+     *# verify that the selected NamedGroup in the ServerHello is the same as
+     *# that in the HelloRetryRequest. If this check fails, the client MUST
+     *# abort the handshake with an "illegal_parameter" alert.
+     **/
+    if (s2n_is_hello_retry_handshake(conn) && !s2n_is_hello_retry_message(conn)) {
+        POSIX_ENSURE_REF(server_ecc_evp_params->negotiated_curve);
+        POSIX_ENSURE(ecc_pref->ecc_curves[supported_curve_index] == server_ecc_evp_params->negotiated_curve,
+                     S2N_ERR_BAD_MESSAGE);
+    }
+
     server_ecc_evp_params->negotiated_curve = ecc_pref->ecc_curves[supported_curve_index];
```

Comment:
This is a little unclear; I had to read it a couple time. You might want to create one or two named variables to make it clear that "ecc_pref->ecc_curves[supported_curve_index]" is the new negotiated curve and "server_ecc_evp_params->negotiated_curve" is the old one.

Reply:
I tried to make this part clearer in ea0a1f9